src/protocol/sender.rs

```diff
@@ -0,0 +1,622 @@
+}
+
+impl LossList {
+    pub fn new(_settings: &ConnectionSettings) -> Self {
+        Self {
+            list: VecDeque::new(),
+        }
+    }
+
+    pub fn push_back(&mut self, packet: DataPacket) {
+        self.list.push_back(packet);
+    }
+
+    pub fn pop_front(&mut self) -> Option<DataPacket> {
+        self.list.pop_front()
+    }
+
+    pub fn remoeve_acknowledged_packets(&mut self, acknowledged: SeqNumber) -> u32 {
+        let mut retransmited_packets = 0;
+        while let Some(x) = self.list.front() {
+            if acknowledged > x.seq_number {
+                let _ = self.pop_front();
+                // this means a packet was lost then retransmitted
+                retransmited_packets += 1;
+            } else {
+                break;
+            }
+        }
+        retransmited_packets
+    }
+
+    pub fn back(&self) -> Option<&DataPacket> {
+        self.list.back()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.list.is_empty()
+    }
+}
+
+#[derive(Debug)]
+pub enum SenderError {}
+
+pub type SenderResult = Result<Option<(ControlPacket, SocketAddr)>, SenderError>;
+
+pub struct SenderMetrics {
+    /// Round trip time, in microseconds
+    pub rtt: i32,
+
+    /// Round trip time variance
+    pub rtt_var: i32,
+
+    /// packet arrival rate
+    pub pkt_arr_rate: i32,
+
+    /// estimated link capacity
+    pub est_link_cap: i32,
+
+    /// Total lost packets
+    pub lost_packets: u32,
+
+    /// Total retransmitted packets
+    pub retrans_packets: u32,
+
+    /// Total received packets (packets that have been ACKed)
+    pub recvd_packets: u32,
+}
+
+impl SenderMetrics {
+    pub fn new() -> Self {
+        Self {
+            rtt: 10_000,
+            rtt_var: 0,
+            pkt_arr_rate: 0,
+            est_link_cap: 0,
+            lost_packets: 0,
+            retrans_packets: 0,
+            recvd_packets: 0,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum SenderAlgorithmStep {
+    Step1,
+    Step2(Instant),
+    Step3(Instant),
+    Step4(Instant),
+    Step5(Instant),
+    Step6(Instant),
+}
+
+#[derive(Debug, Clone)]
+pub enum SenderAlgorithmAction {
+    Continue,
+    WaitUnitlAck,
+    WaitUntilData,
+    WaitUntil(Instant),
+    SendControl((ControlPacket, SocketAddr)),
+    SendData((DataPacket, SocketAddr)),
+    Close,
+}
+
+pub struct Sender {
+    close: bool,
+
+    current_algorithm_step: SenderAlgorithmStep,
+
+    handshake: Handshake,
+
+    /// The settings, including remote sockid and address
+    settings: ConnectionSettings,
+
+    metrics: SenderMetrics,
+
+    /// The buffer to store packets for retransmission, sorted chronologically
+    send_buffer: SendBuffer,
+
+    /// The buffer to store the next control packet
+    control_buffer: Option<ControlPacket>,
+
+    /// The buffer to store packets for transmission
+    transmit_buffer: TransmitBuffer,
+
+    // 1) Sender's Loss List: The sender's loss list is used to store the
```

Comment:
Again, good to have a link to the spec you copied this from (i'm assuming https://tools.ietf.org/html/draft-gg-udt-03)

Reply:
Yes, good point. Though I was following the lead of the existing code, it wasn't consistently provided elsewhere. I'll be sure to add links where suitable from now on.